### PR TITLE
kotatogram-desktop: init at 1.1.5

### PR DESCRIPTION
**Pull request by ilya-fedin**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by balsoft** [comment]
@GrahamcOfBorg build kotatogram-desktop libtgvoip rlottie-tdesktop

**Comment by balsoft** [comment]
I'd like to see three commits, because there are three new packages added. Add them in the order of dependency -> dependent.

**Comment by balsoft** [comment]
Also, if you have a bit of spare time to maintain the packages, please add yourself as a maintainer of these packages.

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (balsoft) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: This looks like something that can be optional. Add `animatedStickers` argument with default value set to `true` and then do `] ++ optional animatedStickers animatedStickersPatch;`

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (balsoft) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Are you sure that this can't be done without setting `NIX_CFLAGS_COMPILE` manually?

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (balsoft) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: You may want to put sources for patches in a separate `let ... in`.

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: animatedStickersPatch is a fix for animated stickers. In k1.1.1 (and tdesktop v1.9.1) there is a bug with flickering stickers

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: I can drop this with my patch for QtDBus-based notifications, so Kotatogram will not needed GTK. I just didn't except that you offer me to being a maintainer.

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Yes. tdesktop build scripts don't support pkgconfig because they are optimized for static compilation.

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (balsoft) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Ah, okay.

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (balsoft) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: I'm still inclined to believe that those `-I` flags shouldn't be passed manually.

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (balsoft) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: I'm not saying you should drop this, all I'm saying is that unneeded arguments to `mkDerivation` are discouraged.

Comment (balsoft) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: I'm not saying you should drop this patch, all I'm saying is that extra unneeded arguments to mkDerivation are discouraged.

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: I cleaned them as much as possible
Compare with https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/telegram/tdesktop/default.nix#L73

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Hmm, ok

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (balsoft) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Ah, interesting. I wonder why this is not handled by `addPackage` in `stdenv/setup`.

Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Because there are no support for pkgconfig in tdesktop build scripts.

**Comment by balsoft** [comment]
cc @primeos  @abbradar

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (primeos) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Probably makes sense to use the following website: https://kotatogram.github.io/

I wasn't aware of Kotatogram and what the differences are. Unfortunately the GitHub repo seems so miss a link to the homepage which makes it difficult to find...

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (primeos) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Would be helpful for users to add something like this:
> It contains some useful (or purely cosmetic) features, but they could be unstable. A detailed list is available here: https://kotatogram.github.io/changes

**Review by primeos — DISMISSED** [review]
This adds a lot of patches and files like `CMakeLists.inj`. Can those be fetched from somewhere (e.g. the AUR)? AFAIK fetching patches is the preferred solution (keeps Nixpkgs cleaner, is easier to review, the source is obvious, etc.).

The rest seems ok from a quick look. Thanks for your contribution :)

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Fixed

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Fixed

**Comment by ilya-fedin** [comment]
> This adds a lot of patches and files like `CMakeLists.inj`. Can those be fetched from somewhere (e.g. the AUR)?

No, preston significantly changed the directory tree in tdesktop v1.9.1, on which kotatogram-desktop v1.1.1 is based. I had to rewrite the patches almost from scratch (but I found a lot of unnecessary changes, most likely these patches were taken from flatpak - these changes are really needed there).
Only no-gtk2.patch remained unchanged.

**Comment by ilya-fedin** [comment]
Updated to k1.1.3 based on v1.9.4

**Review by primeos — COMMENTED** [review]
Seems a lot better now with way fewer patches :)

However, there are still quite a few patches left and I'm not sure what our policy is in such cases (personally I try to avoid it, but of course this isn't always possible). Might be a good idea to cc some very active PR reviewers, as they have way more experience with such cases than I have (e.g. `@jonringer` and `@worldofpeace` might be good fits). Or to mention this PR on Discourse, e.g. in this thread: https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/.

**Comment by ilya-fedin** [comment]
cc @jonringer @worldofpeace

**Review by jonringer — COMMENTED** [review]
I feel like a lot of these patches could be upstreamed, and are not specific to nix

**Comment by ilya-fedin** [comment]
So I should remove them?

**Comment by jonringer** [comment]
By upstreaming I mean, theses could be submitted to the package repos themselves. Unless it's a nix-specific patch, then they should stay here. But many of these do not seem to be specific to NixOS/nixpkgs

**Comment by ilya-fedin** [comment]
What do you mean by "package repos"? nixpkgs is a package repo, does not?
If you mean that I should submit them directly to the software developers, then this is in progress
But question is what I should do with them in the derivation

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/Use-system-font-by-default.patch
Comment (jonringer) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/Use-system-font-by-default.patch]: This feels useful to the kotatogram-desktop repository, A PR should probably be made against it

**Comment by jonringer** [comment]
> What do you mean by "package repos"? nixpkgs is a package repo, does not?
> If you mean that I should submit them directly to the software developers, then this is in progress
> But question is what I should do with them in the derivation

Once the upstream package (such as kaotatogram-desktop) has merged the PRs relating to the patches, then just package the messenger as normal. But then you don't have to apply patches

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/Use-system-font-by-default.patch
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/Use-system-font-by-default.patch]: This setting introduced by other patch, I just separate it for convenience

Comment (jonringer) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/Use-system-font-by-default.patch]: I noticed. But I feel like all the patches should be made into a PR for the kotatogram package, as that's probably what most Linux users want.

What I'm trying to say is, doing patches to the nix packages works, but if it's able to benefit the greater community, then the patches should go into master on the related project

Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/Use-system-font-by-default.patch]: They will, but at that point there are only this way

**Comment by ilya-fedin** [comment]
Okay, I cherry-picked the pkg-config from libtgvoip upstream and removed a couple of minor patches. I left system-tray-icon.patch, because I really want people to be able to have an icon from the system theme in the tray.

**Comment by ilya-fedin** [comment]
@GrahamcOfBorg build kotatogram-desktop libtgvoip rlottie-tdesktop

**Comment by balsoft** [comment]
Why does this fail on aarch64? Is it known upstream that this fails?

**Comment by ilya-fedin** [comment]
Fails not this package, but it's dependency - range-v3. It's fails on tests compilation.

**Comment by ilya-fedin** [comment]
Updated to k1.1.4 based on v1.9.7 and dropped all the patches

**Review by primeos — COMMENTED** [review]
Looks pretty good to me now, added some recommendations but would like a second opinion, mainly regarding the naming in `all-packages.nix` / the additional two packages in general.

**Comment by jonringer** [comment]
@GrahamcOfBorg build kotatogram-desktop
@GrahamcOfBorg build libtgvoip
@GrahamcOfBorg build rlottie-tdesktop

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (jonringer) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: if you don't want to troubleshoot aarch issues
```suggestion
    platforms = platforms.linux;
    badPlatforms = platforms.aarch64;
```

Comment (jonringer) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: issue seems to be an upstream issue:
```
/build/source/test/view/set_symmetric_difference.cpp:324:1: internal compiler error: in output_constructor_regular_field, at varasm.c:5207
  324 | }
      | ^
```

Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: > if you don't want to troubleshoot aarch issues

But issue is in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/range-v3/default.nix
This can be fixed if maintainer of range-v3 disablse tests for aarch64

File: pkgs/top-level/all-packages.nix
Comment (ilya-fedin) [reply pkgs/top-level/all-packages.nix]: libtgvoip upstream is dead and tdesktop's libtgvoip is the new upstream de-facto
I can enable bundling of rlottie

File: pkgs/top-level/all-packages.nix
Comment (ilya-fedin) [reply pkgs/top-level/all-packages.nix]: fixed

File: pkgs/development/libraries/libtgvoip/default.nix
Comment (ilya-fedin) [reply pkgs/development/libraries/libtgvoip/default.nix]: fixed

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: fixed

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (primeos) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Right, I'll disable it in https://github.com/NixOS/nixpkgs/pull/78447 (was aware of it but initially hoping that it'll be fixed soon, but I didn't find an upstream fix and don't have an AArch64 test setup atm - theoretically it should work though, at least the Alpine Linux package builds).

Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: Why just not to disable doCheck for aarch64?
The library is header-only and errors are from tests

Comment (jonringer) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: if you can verify that it's only the tests which are affected, then that's fine. Just worried about the instance where the build is actually affected.

Comment (primeos) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: > Why just not to disable doCheck for aarch64?
The library is header-only and errors are from tests

I wasn't sure it the errors are actually from the tests (didn't have a closer look and I'm used that the tests run after the actual build - like with Makefiles or Python). Let's see if it works, if not I'll just disable it until someone can properly test and fix it.

And please add further comments on the other PR if we don't add the commit in this PR.

**Comment by ilya-fedin** [comment]
@GrahamcOfBorg build kotatogram-desktop libtgvoip

**Comment by ilya-fedin** [comment]
updated to k1.1.5 based on v1.9.8

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (jonringer) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: please use canonical name
```suggestion
  nativeBuildInputs = [ pkg-config python3 cmake ninja ];
```

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (jonringer) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]:
```suggestion
{ mkDerivation, lib, fetchFromGitHub, pkg-config, python3, cmake, ninja
```

File: pkgs/development/libraries/libtgvoip/default.nix
Comment (jonringer) [inline pkgs/development/libraries/libtgvoip/default.nix]:
```suggestion
  nativeBuildInputs = [ pkg-config autoreconfHook ];
```

File: pkgs/development/libraries/libtgvoip/default.nix
Comment (jonringer) [inline pkgs/development/libraries/libtgvoip/default.nix]:
```suggestion
{ stdenv, lib, fetchFromGitHub, pkg-config, autoreconfHook
```

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: fixed

File: pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]:
```suggestion
```

this is default enabled in cmake and ninja setup hooks

Comment (ilya-fedin) [reply pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix]: fixed

**Review by jonringer — APPROVED** [review]
diff LGTM
application launches fine

```
[5 built, 88 copied (131.9 MiB), 24.3 MiB DL]
https://github.com/NixOS/nixpkgs/pull/75210
2 package built:
kotatogram-desktop libtgvoip
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build kotatogram-desktop libtgvoip

**Review by balsoft — APPROVED** [review]

**Comment by jonringer** [comment]
aarch failure unrelated

**Pull request merged**
